### PR TITLE
Prevent flux from accidentally deleting serviceaccount when using ControllerConfig

### LINKED ISSUE #3030: Flux compatibility - service account annotations
### What happened?
We sometimes see bursts of Unauthorized for calls from crossplane providers to the kubernetes apiserver. After some debugging we've found that the UID in the service account does not match the token on the file system. We noticed that the service account has a "fluxcd.io" annotation, even though it is not managed by flux. This is because crossplane copies all annotations from the `ControllerConfig` to the `ServiceAccount` object. This is probably required for ArgoCD, and it is used for IRSA in AWS.

But for flux, this does not work. We believe flux will consider this as a situation where a former `Kustomization` managed this object, and delete it as it is no longer managed. crossplane will quickly recreate it, but the token mounted in the container will for some time have the old UID and the API server will reject the requests. I'm not sure if this is a race condition in the kubelet when a token is rotated while a pod is created, or if it is simply not a supported way to rotate the tokens by deleting the service account while it is in use, but Kubernetes does recover from this situation after a while.

The specific annotation is `"kustomize.toolkit.fluxcd.io/checksum": "3583e44b78adb421bd53fcb43edfcd5d39b287b8"`

### How can we reproduce it?
Flux-managed `ControllerConfig`, see above

### Suggested fix

I think we should exclude annotations with keys `fluxcd.io/` or prefix `kustomize.toolkit.fluxcd.io/` when creating the deployment and serviceaccount in

https://github.com/crossplane/crossplane/blob/1344a86018f03fcbba6b9365c63696dcd47ebcf6/internal/controller/pkg/revision/deployment.go#L174-L177

We might also want to skip key `kubectl.kubernetes.io/last-applied-configuration` or prefix `kubectl.kubernetes.io/` as it is fairly misleading to copy it.

Long term crossplane should probably implement "bring your own service account" instead (see issue https://github.com/crossplane/crossplane/issues/2295 etc)

### What environment did it happen in?
Crossplane version: 1.6.3, EKS 1.21, flux 0.17

### PR DESCRIPTION
**Pull request by chlunde**
Flux will garbage collect the ServiceAccount if we copy the flux
annotations from the ControllerConfig, when the ControllerConfig is managed by flux and the ServiceAccount is not.

### Description of your changes

Exclude known bad and meaningless annotations when copying annotations.

Fixes #3030

I have:

- [x] Read and followed Crossplane's [contribution process].
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

Unit tests

[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Review by jjcallis — APPROVED** [review]

**Review by haarchri — APPROVED** [review]
tested this in our flux environment - works as expected now - LGTM

File: internal/controller/pkg/revision/deployment.go
Comment (negz) [inline internal/controller/pkg/revision/deployment.go:281]: Does it make sense to make these a "constant" (i.e. a `var doNotPropagate []string = ...`)?

Comment (negz) [reply internal/controller/pkg/revision/deployment.go:281]: Following up here. I think this is mergeable if you have time to address this. :)

Comment (chlunde) [reply internal/controller/pkg/revision/deployment.go:281]: @negz sorry, busy times! doNotPropagate at function level, or package?

Comment (negz) [reply internal/controller/pkg/revision/deployment.go:281]: I was thinking package (sorry also for the huge delay in response times).

Comment (chlunde) [reply internal/controller/pkg/revision/deployment.go]: @negz At least we have a stable cadence :)

**Comment by wwentland** [comment]
It would be fantastic if this could be reviewed again! I believe that @negz suggestion has been addressed.

**Review by nullable-eth — APPROVED** [review]
Agree that @negz comment was addressed, no other objections.

**Comment by chlunde** [comment]
@nullable-eth FYI I don't have a merge button, a maintainer must merge

**Comment by nullable-eth** [comment]
> @nullable-eth FYI I don't have a merge button, a maintainer must merge

I think this still requires another code owner to review.

**Comment by wwentland** [comment]
> > @nullable-eth FYI I don't have a merge button, a maintainer must merge
> 
> I think this still requires another code owner to review.

@nullable-eth Thank you! Who might be best placed to help us with that?

**Review by turkenh — APPROVED** [review]
Looking good to me, but needs a rebase before merge!

**Review by hasheddan — COMMENTED** [review]
I am currently hesitant to continue to evolve this set of functionality (i.e. Provider deployment and configuration) given a desire to move towards customer provider runtimes as defined in https://github.com/crossplane/crossplane/issues/2671. In that model, our current functionality of creating a `Deployment` with certain attributes would be moved into a separate component, where folks could customize and or alter the behavior / level of customization as they see fit. @chlunde would you mind taking a peek at that proposal and weighing in if you have any thoughts?

**Comment by negz** [comment]
Should we close this PR?

**Pull request closed without merging**
